File: src/ServerDatabaseIdentity.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace Clipman
{
    internal static class ServerDatabaseIdentity
    {
        private const string Purpose = "Clipman.ServerDatabaseId.v1";
        private const string ChannelPurpose = "Clipman.ServerChannelId.v1";
        private const string SyncRulesPurpose = "Clipman.ServerSyncRulesId.v1";

        public static string FromTokenAndPassword(string serverToken, string historyPassword)
        {
            var password = historyPassword ?? string.Empty;
            return Derive(serverToken, password, Purpose + "\n" + password);
        }

        public static string ChannelFromTokenAndPassword(string serverToken, string historyPassword, string channelKey)
        {
            var password = historyPassword ?? string.Empty;
            return Derive(serverToken, password, ChannelPurpose + "\n" + password + "\n" + (channelKey ?? string.Empty));
        }

        public static string SyncRulesFromTokenAndPassword(string serverToken, string historyPassword)
        {
            var password = historyPassword ?? string.Empty;
            return Derive(serverToken, password, SyncRulesPurpose + "\n" + password);
        }

        private static string Derive(string serverToken, string password, string message)
        {
            var token = (serverToken ?? string.Empty).Trim();
            if (token.Length == 0 || password.Length == 0) return string.Empty;

            var key = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(token));
            using (var hmac = new HMACSHA256(key))
            {
                return ToBase64Url(hmac.ComputeHash(Encoding.UTF8.GetBytes(message)));
            }
        }

        private static string ToBase64Url(byte[] bytes)
        {
            return Convert.ToBase64String(bytes ?? new byte[0]).TrimEnd('=').Replace('+', '-').Replace('/', '_');
        }
    }
}

File: ClipmanAndroid/app/src/main/java/me/onj/clipman/ServerDatabaseIdentity.kt
package me.onj.clipman

import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import java.security.MessageDigest

/**
 * The bucket identity derivations of sync-rules-spec.md section 2. Every id is
 * base64url without padding over HMAC-SHA256 keyed with SHA-256 of the trimmed
 * server token, and a blank token or password yields an empty id.
 */
object ServerDatabaseIdentity {
    private const val purpose = "Clipman.ServerDatabaseId.v1"
    private const val channelPurpose = "Clipman.ServerChannelId.v1"
    private const val syncRulesPurpose = "Clipman.ServerSyncRulesId.v1"

    fun fromTokenAndPassword(serverToken: String, historyPassword: String): String =
        derive(serverToken, historyPassword, "$purpose\n$historyPassword")

    /**
     * The bucket that holds one sync channel. The channel key is normalized
     * first, so an unroutable channel name addresses no bucket.
     */
    fun channelId(serverToken: String, historyPassword: String, channelKey: String): String {
        val key = SyncRuleEngine.channelKey(channelKey)
        if (key.isEmpty()) return ""
        return derive(serverToken, historyPassword, "$channelPurpose\n$historyPassword\n$key")
    }

    /** The bucket that holds the sync rules document. */
    fun syncRulesId(serverToken: String, historyPassword: String): String =
        derive(serverToken, historyPassword, "$syncRulesPurpose\n$historyPassword")

    private fun derive(serverToken: String, historyPassword: String, message: String): String {
        val token = serverToken.trim()
        if (token.isEmpty() || historyPassword.isEmpty()) return ""
        val key = MessageDigest.getInstance("SHA-256").digest(token.toByteArray(Charsets.UTF_8))
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        val digest = mac.doFinal(message.toByteArray(Charsets.UTF_8))
        return Base64.getUrlEncoder().withoutPadding().encodeToString(digest)
    }
}
